data/tabular/scaffold_split.py

```diff
@@ -0,0 +1,182 @@
+    "choline_transporter_butkiewicz",
+    "orexin1_receptor_butkiewicz",
+    "human_intestinal_absorption",
+    "nr_ahr_tox21",
+    "cyp3a4_substrate_carbonmangels",
+    "herg_karim_et_al",
+    "hiv",
+    "carcinogens",
+    "sr_are_tox21",
+    "nr_ppar_gamma_tox21",
+    "solubility_aqsoldb",
+    "m1_muscarinic_receptor_agonists_butkiewicz",
+    "ames_mutagenicity",
+    "potassium_ion_channel_kir2_1_butkiewicz",
+    "cyp_p450_3a4_inhibition_veith_et_al",
+    "skin_reaction",
+    "cyp2d6_substrate_carbonmangels",
+    "cyp_p450_2c19_inhibition_veith_et_al",
+    "nr_ar_lbd_tox21",
+    "p_glycoprotein_inhibition_broccatelli_et_al",
+    "bioavailability_ma_et_al",
+    "BACE",
+    "hiv",
+    "lipophilicity",
+    "thermosol",
+    "MUV_466",
+    "MUV_548",
+    "MUV_600",
+    "MUV_644",
+    "MUV_652",
+    "MUV_689",
+    "MUV_692",
+    "MUV_712",
+    "MUV_713",
+    "MUV_733",
+    "MUV_737",
+    "MUV_810",
+    "MUV_832",
+    "MUV_846",
+    "MUV_852",
+    "MUV_858",
+    "MUV_859",
+]
+
+
+def split_for_smiles(smiles: str, train_smiles: List[str], val_smiles: List[str]):
+    if smiles in train_smiles:
+        return "train"
+    elif smiles in val_smiles:
+        return "valid"
+    else:
+        return "test"
+
+
+def main(
+    data_dir,
+    override: bool = False,
+    train_frac: float = 0.7,
+    val_frac: float = 0.15,
+    test_frac: float = 0.15,
+    seed: int = 42,
+    debug: bool = False,
+):
+    all_yaml_files = glob(os.path.join(data_dir, "**", "*.yaml"), recursive=True)
+    if debug:
+        all_yaml_files = all_yaml_files[:5]
+    transformed_files = []
+    for file in all_yaml_files:
+        with open(file, "r") as f:
+            meta = yaml.safe_load(f)
+
+        if meta["name"] in to_scaffold_split:
+            # run transform.py script in this directory if `data_clean.csv` does not exist
+            # use this dir as cwd
+            if not os.path.exists(
+                os.path.join(os.path.dirname(file), "data_clean.csv")
+            ):
+                subprocess.run(
+                    ["python", "transform.py"],
+                    cwd=os.path.dirname(file),
+                )
+
+            transformed_files.append(
+                os.path.join(os.path.dirname(file), "data_clean.csv")
+            )
+
+    all_smiles = set()
+    for file in transformed_files:
+        df = pd.read_csv(file)
+        all_smiles.update(df["SMILES"].tolist())
+        del df  # ensure memory is freed
+
+    all_smiles = list(all_smiles)
+    splits = _create_scaffold_split(
+        all_smiles, frac=[train_frac, val_frac, test_frac], seed=42
+    )
+
+    # select the right indices for each split
+    train_smiles = [all_smiles[i] for i in splits["train"]]
+    val_smiles = [all_smiles[i] for i in splits["valid"]]
+    print(
+        "Train smiles:",
+        len(train_smiles),
+        "Val smiles:",
+        len(val_smiles),
+        "Test smiles:",
+        len(splits["test"]),
```

Comment:
why no `test_smiles` here and above?

Reply:
cause we do not allocate a list of smiles for test (we only have the indices in the dict)